fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -0,0 +1,198 @@
+    heuristics to find your own optimal configuration.
+
+    Args:
+        module (torch.nn.Sequential):
+            sequential module to be parallelized
+        balance (ints):
+            list of number of layers in each partition
+
+    Keyword Args:
+        devices (iterable of devices):
+            devices to use (default: all CUDA devices)
+        chunks (int):
+            number of micro-batches (default: ``1``)
+        checkpoint (str):
+            when to enable checkpointing, one of ``'always'``,
+            ``'except_last'``, or ``'never'`` (default: ``'except_last'``)
+        deferred_batch_norm (bool):
+            whether to use deferred BatchNorm moving statistics (default:
+            :data:`False`, see :class:`Deferred Batch Normalization <DeferredBatchNorm>` for more
+            details)
+
+    Raises:
+        TypeError:
+            the module is not a :class:`nn.Sequential <torch.nn.Sequential>`.
+        ValueError:
+            invalid arguments, or wrong balance
+        IndexError:
+            the number of devices is fewer than the number of partitions.
+
+    """
+
+    #: The number of micro-batches.
+    chunks: int = 1
+
+    #: The checkpoint mode to determine when to enable checkpointing. It is one
+    #: of ``'always'``, ``'except_last'``, or ``'never'``.
+    checkpoint: str = "never"
+
+    def __init__(
+        self,
+        module: List[LayerSpec],
+        *,
+        balance: List[int],
+        devices: List[str],
+        chunks: int = chunks,
+        checkpoint: str = checkpoint,
+        deferred_batch_norm: bool = False,
+    ) -> None:
+        super().__init__()
+
+        if type(chunks) is not int or chunks <= 0:
+            raise ValueError("number of chunks must be positive integer")
+        if checkpoint not in ["never"]:
+            raise ValueError("checkpoint is not yet implemented")
+        if deferred_batch_norm:
+            raise ValueError("deferred_batch_norm is not yet implemented")
+        if len(balance) != len(devices):
+            raise ValueError("balance and devices lists must be the same size")
+        if len(module) != sum(balance):
+            raise ValueError("number of layers must match aggregate balance")
+
+        _verify_module(module)
+
+        index = 0
+        rmodule = []
+        workers = []
+        for num_layers, device_spec in zip(balance, devices):
+            worker, device = device_spec.split("/")
+            next_index = index + num_layers
+            rlayer = rpc.remote(worker, _create_sequential, args=(module[index:next_index], device))
+            index = next_index
+            workers.append(worker)
+            rmodule.append(rlayer)
+
+        self.chunks = chunks
+        self.checkpoint = checkpoint
+        self.module = module
+        self.workers = workers
+        self.rmodule = rmodule
+
+    def forward(self, x: Tensor) -> rpc.RRef:  # type: ignore
```

Comment:
If i am understanding correctly, this forward() impl doesn't support checkpointing right?

Reply:
It does not yet although it may support PyTorch's regular checkpointing. I will test that.